### PR TITLE
JAMES-3763 Generics for BlobIdProvider

### PR DESCRIPTION
**Pull request by chibenwa**
It avoids buffering byte arrays onto temporary files..

Triggering IOs is a MAJOR performance hit in itself but adding this onto a dedicated large pool for IOs when we are compute bound also skyrockets the count of threads. Not to mention that my experience with resource management with reactor let me think that edge cases like cancelation could trigger a ressource leak (been there).

We can beatifully avoid these inneficiencies by making BlobIdProvider a generic.

Disaster that waited to happen...

### CODE REVIEW

File: server/blob/blob-storage-strategy/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala [hunk 1]
```diff
@@ -67,40 +67,35 @@ class DeDuplicationBlobStore @Inject()(blobStoreDAO: BlobStoreDAO,
...
-    save(bucketName, data, withBlobId, storagePolicy)
+    save(bucketName, data, withBlobIdFromByteSource, storagePolicy)
   }
 
-  override def save(bucketName: BucketName, data: Array[Byte], blobIdProvider: BlobIdProvider, storagePolicy: BlobStore.StoragePolicy): Publisher[BlobId] = {
+  override def save(bucketName: BucketName, data: Array[Byte], blobIdProvider: BlobIdProvider[Array[Byte]], storagePolicy: BlobStore.StoragePolicy): Publisher[BlobId] = {
     Preconditions.checkNotNull(bucketName)
     Preconditions.checkNotNull(data)
-    save(bucketName, new ByteArrayInputStream(data), blobIdProvider, storagePolicy)
+    save(bucketName, data, blobIdProvider, storagePolicy)
```
Comment (quantranhong1999) [inline server/blob/blob-storage-strategy/src/main/scala/org/apache/james/server/blob/deduplication/DeDuplicationBlobStore.scala:84]: Likely not finished here

```suggestion
    SMono(blobIdProvider.apply(data))
      .map(_.getT1)
      .flatMap(blobId => SMono(blobStoreDAO.save(bucketName, blobId, data))
      .`then`(SMono.just(blobId)))
```

Should solve the `java.lang.StackOverflowError`. We should double check other places too cc @vttranlina.

**Comment by Arsnael** [comment]
Same issue as reported earlier by Quan regarding passthrough blobstore : https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-2540/2/#showFailuresLink

Should be fixed in the fixup

**Comment by quantranhong1999** [comment]
Note: perf test should be executed

**Review by Arsnael — APPROVED** [review]

**Comment by vttranlina** [comment]
> Note: perf test should be executed

I'm doing it

**Review by chibenwa — COMMENTED** [review]
I'm ok with latest fixups.
Thanks

**Comment by vttranlina** [comment]
## Performance test
The 99th percentile metric for `blobStore::saveBytes` is slightly degraded
Gatling result for JMAP `submitEmails` too.


### Metric
- Before
```
blobStore:saveBytes{quantile="0.5",} 0.06920601500000001
blobStore:saveBytes{quantile="0.75",} 0.161480703
blobStore:saveBytes{quantile="0.95",} 0.327155711
blobStore:saveBytes{quantile="0.98",} 0.528482303
blobStore:saveBytes{quantile="0.99",} 0.683671551
blobStore:saveBytes{quantile="0.999",} 1.9545456630000002
blobStore:saveBytes_count 118781.0
```
- After
```
blobStore:saveBytes{quantile="0.5",} 0.074448895
blobStore:saveBytes{quantile="0.75",} 0.16462643100000002
blobStore:saveBytes{quantile="0.95",} 0.352321535
blobStore:saveBytes{quantile="0.98",} 0.578813951
blobStore:saveBytes{quantile="0.99",} 0.750780415
blobStore:saveBytes{quantile="0.999",} 2.365587455
blobStore:saveBytes_count 120382.0
```


### Gatling 
Screenshot 2024-12-03 at 9 32 50 PM

Screenshot 2024-12-03 at 9 34 23 PM

**Comment by chibenwa** [comment]
@vttranlina The gatling simulation sends mostly very short email

This enhancement is function of the size of the blob so for uuid-sized blobs it is not surprising that it is unrelevant.

I think of `provisionning` as a way more relavant way to triage this as it creates messages with latrger size variations.

Can you please take the time to run provisionning 10 minutes twice tomorrow morning?

Thanks by advance.

**Comment by chibenwa** [comment]
image

Small blobs are held in memory. Thus with blobs of 10KB enhancements are not met.

Also previous version had the good idea to do hashing off-band on a dedicated thread and not potentially on a driver thread. I opened a follow up draft PR switching hasing to parallel thread cf https://github.com/apache/james-project/pull/2544

@vttranlina  Can you please run provisionning on:

 - Before
 - This version
 - https://github.com/apache/james-project/pull/2544

**Pull request merged**